File: go/strictcli/tagdsl.go
package strictcli

import (
	"fmt"
	"strings"
)

// tagTokenType represents the type of a token in a tag expression.
type tagTokenType int

const (
	tagTokenIdent  tagTokenType = iota // identifier (tag name)
	tagTokenAnd                        // &
	tagTokenOr                         // |
	tagTokenXor                        // ^
	tagTokenDiff                       // -
	tagTokenNot                        // !
	tagTokenLParen                     // (
	tagTokenRParen                     // )
)

// tagToken is a single token from a tag expression.
type tagToken struct {
	typ tagTokenType
	val string
	pos int
}

// tagNode is a node in the tag expression AST.
type tagNode interface {
	eval(tags map[string]bool) bool
}

// tagIdent is a leaf node representing a tag name.
type tagIdent struct {
	name string
}

func (n *tagIdent) eval(tags map[string]bool) bool {
	return tags[n.name]
}

// tagNot is a unary NOT node.
type tagNot struct {
	operand tagNode
}

func (n *tagNot) eval(tags map[string]bool) bool {
	return !n.operand.eval(tags)
}

// tagAnd is a binary AND node.
type tagAnd struct {
	left, right tagNode
}

func (n *tagAnd) eval(tags map[string]bool) bool {
	return n.left.eval(tags) && n.right.eval(tags)
}

// tagOr is a binary OR node.
type tagOr struct {
	left, right tagNode
}

func (n *tagOr) eval(tags map[string]bool) bool {
	return n.left.eval(tags) || n.right.eval(tags)
}

// tagXor is a binary XOR node.
type tagXor struct {
	left, right tagNode
}

func (n *tagXor) eval(tags map[string]bool) bool {
	return n.left.eval(tags) != n.right.eval(tags)
}

// tagDiff is a binary DIFF node (left AND NOT right).
type tagDiff struct {
	left, right tagNode
}

func (n *tagDiff) eval(tags map[string]bool) bool {
	return n.left.eval(tags) && !n.right.eval(tags)
}

// tagdslTokenize splits a tag expression string into tokens.
func tagdslTokenize(expr string) ([]tagToken, error) {
	var tokens []tagToken
	i := 0
	for i < len(expr) {
		ch := expr[i]

		// Skip whitespace
		if ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r' {
			i++
			continue
		}

		switch ch {
		case '&':
			tokens = append(tokens, tagToken{typ: tagTokenAnd, val: "&", pos: i})
			i++
		case '|':
			tokens = append(tokens, tagToken{typ: tagTokenOr, val: "|", pos: i})
			i++
		case '^':
			tokens = append(tokens, tagToken{typ: tagTokenXor, val: "^", pos: i})
			i++
		case '-':
			tokens = append(tokens, tagToken{typ: tagTokenDiff, val: "-", pos: i})
			i++
		case '!':
			tokens = append(tokens, tagToken{typ: tagTokenNot, val: "!", pos: i})
			i++
		case '(':
			tokens = append(tokens, tagToken{typ: tagTokenLParen, val: "(", pos: i})
			i++
		case ')':
			tokens = append(tokens, tagToken{typ: tagTokenRParen, val: ")", pos: i})
			i++
		default:
			// Try to read an identifier: [a-z][a-z0-9-]*
			if ch >= 'a' && ch <= 'z' {
				start := i
				i++
				for i < len(expr) {
					c := expr[i]
					if (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') || c == '-' {
						i++
					} else {
						break
					}
				}
				tokens = append(tokens, tagToken{typ: tagTokenIdent, val: expr[start:i], pos: start})
			} else {
				return nil, fmt.Errorf("tag expression: unexpected character %q at position %d", string(ch), i)
			}
		}
	}
	return tokens, nil
}

// tagdslParser holds state during recursive descent parsing.
type tagdslParser struct {
	tokens []tagToken
	pos    int
}

// tagdslParse parses a token list into an AST using recursive descent.
// Precedence (tightest first): NOT, AND, XOR, OR, DIFF.
func tagdslParse(tokens []tagToken) (tagNode, error) {
	if len(tokens) == 0 {
		return nil, fmt.Errorf("tag expression: empty expression")
	}
	p := &tagdslParser{tokens: tokens, pos: 0}
	node, err := p.parseDiff()
	if err != nil {
		return nil, err
	}
	if p.pos < len(p.tokens) {
		tok := p.tokens[p.pos]
		return nil, fmt.Errorf("tag expression: unexpected token %q at position %d", tok.val, tok.pos)
	}
	return node, nil
}

// peek returns the current token without consuming it, or nil if at end.
func (p *tagdslParser) peek() *tagToken {
	if p.pos >= len(p.tokens) {
		return nil
	}
	return &p.tokens[p.pos]
}

// advance consumes and returns the current token.
func (p *tagdslParser) advance() tagToken {
	tok := p.tokens[p.pos]
	p.pos++
	return tok
}

// parseDiff handles the lowest-precedence operator: DIFF (-)
func (p *tagdslParser) parseDiff() (tagNode, error) {
	left, err := p.parseOr()
	if err != nil {
		return nil, err
	}
	for {
		tok := p.peek()
		if tok == nil || tok.typ != tagTokenDiff {
			break
		}
		p.advance()
		right, err := p.parseOr()
		if err != nil {
			return nil, err
		}
		left = &tagDiff{left: left, right: right}
	}
	return left, nil
}

// parseOr handles OR (|)
func (p *tagdslParser) parseOr() (tagNode, error) {
	left, err := p.parseXor()
	if err != nil {
		return nil, err
	}
	for {
		tok := p.peek()
		if tok == nil || tok.typ != tagTokenOr {
			break
		}
		p.advance()
		right, err := p.parseXor()
		if err != nil {
			return nil, err
		}
		left = &tagOr{left: left, right: right}
	}
	return left, nil
}

// parseXor handles XOR (^)
func (p *tagdslParser) parseXor() (tagNode, error) {
	left, err := p.parseAnd()
	if err != nil {
		return nil, err
	}
	for {
		tok := p.peek()
		if tok == nil || tok.typ != tagTokenXor {
			break
		}
		p.advance()
		right, err := p.parseAnd()
		if err != nil {
			return nil, err
		}
		left = &tagXor{left: left, right: right}
	}
	return left, nil
}

// parseAnd handles AND (&)
func (p *tagdslParser) parseAnd() (tagNode, error) {
	left, err := p.parseNot()
	if err != nil {
		return nil, err
	}
	for {
		tok := p.peek()
		if tok == nil || tok.typ != tagTokenAnd {
			break
		}
		p.advance()
		right, err := p.parseNot()
		if err != nil {
			return nil, err
		}
		left = &tagAnd{left: left, right: right}
	}
	return left, nil
}

// parseNot handles NOT (!) as a prefix unary operator.
func (p *tagdslParser) parseNot() (tagNode, error) {
	tok := p.peek()
	if tok != nil && tok.typ == tagTokenNot {
		p.advance()
		operand, err := p.parseNot()
		if err != nil {
			return nil, err
		}
		return &tagNot{operand: operand}, nil
	}
	return p.parsePrimary()
}

// parsePrimary handles identifiers and parenthesized expressions.
func (p *tagdslParser) parsePrimary() (tagNode, error) {
	tok := p.peek()
	if tok == nil {
		lastPos := 0
		if len(p.tokens) > 0 {
			last := p.tokens[len(p.tokens)-1]
			lastPos = last.pos + len(last.val)
		}
		return nil, fmt.Errorf("tag expression: unexpected end of expression at position %d", lastPos)
	}

	switch tok.typ {
	case tagTokenIdent:
		p.advance()
		return &tagIdent{name: tok.val}, nil
	case tagTokenLParen:
		p.advance()
		node, err := p.parseDiff()
		if err != nil {
			return nil, err
		}
		closing := p.peek()
		if closing == nil || closing.typ != tagTokenRParen {
			return nil, fmt.Errorf("tag expression: expected \")\" at position %d", p.endPos())
		}
		p.advance()
		return node, nil
	default:
		return nil, fmt.Errorf("tag expression: unexpected token %q at position %d", tok.val, tok.pos)
	}
}

// endPos returns the position after the last token, for error messages.
func (p *tagdslParser) endPos() int {
	if len(p.tokens) == 0 {
		return 0
	}
	last := p.tokens[len(p.tokens)-1]
	return last.pos + len(last.val)
}

// tagdslEvaluate evaluates an AST node against a set of tags.
func tagdslEvaluate(node tagNode, tags map[string]bool) bool {
	return node.eval(tags)
}

// matchTagExpr tokenizes, parses, and evaluates a tag expression against a tag set.
func matchTagExpr(expr string, tags map[string]bool) (bool, error) {
	expr = strings.TrimSpace(expr)
	if expr == "" {
		return false, fmt.Errorf("tag expression: empty expression")
	}
	tokens, err := tagdslTokenize(expr)
	if err != nil {
		return false, err
	}
	if len(tokens) == 0 {
		return false, fmt.Errorf("tag expression: empty expression")
	}
	node, err := tagdslParse(tokens)
	if err != nil {
		return false, err
	}
	return tagdslEvaluate(node, tags), nil
}

File: typescript/src/checks/tagdsl.ts
/**
 * Tag DSL: set-operation expressions over check tags. Operators by precedence
 * (tightest first): ! (NOT), & (AND), ^ (XOR), | (OR), - (DIFF). Parentheses
 * group. Error positions are byte offsets into the expression string.
 *
 * Parity sources: go/strictcli/tagdsl.go and Python _tagdsl_* (~6661-6834).
 * Errors are thrown as plain Error (the Python ValueError analog at run
 * time); callers surface them per the sibling contract.
 */

import {
	errTagExprEmpty,
	errTagExprExpectedRParen,
	errTagExprUnexpectedChar,
	errTagExprUnexpectedEnd,
	errTagExprUnexpectedToken,
} from "../errors.js";

type TagTokenType =
	| "ident"
	| "and"
	| "or"
	| "xor"
	| "diff"
	| "not"
	| "lparen"
	| "rparen";

interface TagToken {
	readonly typ: TagTokenType;
	readonly val: string;
	readonly pos: number;
}

type TagNode =
	| { readonly kind: "ident"; readonly name: string }
	| { readonly kind: "not"; readonly operand: TagNode }
	| {
			readonly kind: "and" | "or" | "xor" | "diff";
			readonly left: TagNode;
			readonly right: TagNode;
	  };

const SINGLE_CHAR_TOKENS: ReadonlyMap<string, TagTokenType> = new Map([
	["&", "and"],
	["|", "or"],
	["^", "xor"],
	["-", "diff"],
	["!", "not"],
	["(", "lparen"],
	[")", "rparen"],
]);

function isIdentStart(ch: string): boolean {
	return ch >= "a" && ch <= "z";
}

function isIdentPart(ch: string): boolean {
	return (ch >= "a" && ch <= "z") || (ch >= "0" && ch <= "9") || ch === "-";
}

function tokenize(expr: string): TagToken[] {
	const tokens: TagToken[] = [];
	let i = 0;
	while (i < expr.length) {
		const ch = expr[i] as string;
		if (ch === " " || ch === "\t" || ch === "\n" || ch === "\r") {
			i++;
			continue;
		}
		const typ = SINGLE_CHAR_TOKENS.get(ch);
		if (typ !== undefined) {
			// A leading identifier character wins over the DIFF operator inside
			// names: identifiers are consumed greedily below, so "-" only
			// tokenizes as DIFF when it does not continue an identifier.
			tokens.push({ typ, val: ch, pos: i });
			i++;
			continue;
		}
		if (isIdentStart(ch)) {
			const start = i;
			i++;
			while (i < expr.length && isIdentPart(expr[i] as string)) {
				i++;
			}
			tokens.push({ typ: "ident", val: expr.slice(start, i), pos: start });
			continue;
		}
		throw new Error(errTagExprUnexpectedChar(ch, i));
	}
	return tokens;
}

/** Recursive-descent parser state over the token list. */
class TagParser {
	private pos = 0;

	constructor(private readonly tokens: readonly TagToken[]) {}

	private peek(): TagToken | undefined {
		return this.tokens[this.pos];
	}

	private advance(): TagToken {
		const tok = this.tokens[this.pos] as TagToken;
		this.pos++;
		return tok;
	}

	private endPos(): number {
		if (this.tokens.length === 0) {
			return 0;
		}
		const last = this.tokens[this.tokens.length - 1] as TagToken;
		return last.pos + last.val.length;
	}

	parse(): TagNode {
		const node = this.parseDiff();
		const tok = this.peek();
		if (tok !== undefined) {
			throw new Error(errTagExprUnexpectedToken(tok.val, tok.pos));
		}
		return node;
	}

	/** DIFF (-) is the loosest-binding operator. */
	private parseDiff(): TagNode {
		let left = this.parseOr();
		while (this.peek()?.typ === "diff") {
			this.advance();
			left = { kind: "diff", left, right: this.parseOr() };
		}
		return left;
	}

	private parseOr(): TagNode {
		let left = this.parseXor();
		while (this.peek()?.typ === "or") {
			this.advance();
			left = { kind: "or", left, right: this.parseXor() };
		}
		return left;
	}

	private parseXor(): TagNode {
		let left = this.parseAnd();
		while (this.peek()?.typ === "xor") {
			this.advance();
			left = { kind: "xor", left, right: this.parseAnd() };
		}
		return left;
	}

	private parseAnd(): TagNode {
		let left = this.parseNot();
		while (this.peek()?.typ === "and") {
			this.advance();
			left = { kind: "and", left, right: this.parseNot() };
		}
		return left;
	}

	/** NOT (!) is a prefix unary operator and binds tightest. */
	private parseNot(): TagNode {
		const tok = this.peek();
		if (tok !== undefined && tok.typ === "not") {
			this.advance();
			return { kind: "not", operand: this.parseNot() };
		}
		return this.parsePrimary();
	}

	private parsePrimary(): TagNode {
		const tok = this.peek();
		if (tok === undefined) {
			throw new Error(errTagExprUnexpectedEnd(this.endPos()));
		}
		if (tok.typ === "ident") {
			this.advance();
			return { kind: "ident", name: tok.val };
		}
		if (tok.typ === "lparen") {
			this.advance();
			const node = this.parseDiff();
			const closing = this.peek();
			if (closing === undefined || closing.typ !== "rparen") {
				throw new Error(errTagExprExpectedRParen(this.endPos()));
			}
			this.advance();
			return node;
		}
		throw new Error(errTagExprUnexpectedToken(tok.val, tok.pos));
	}
}

function evaluate(node: TagNode, tags: ReadonlySet<string>): boolean {
	switch (node.kind) {
		case "ident":
			return tags.has(node.name);
		case "not":
			return !evaluate(node.operand, tags);
		case "and":
			return evaluate(node.left, tags) && evaluate(node.right, tags);
		case "or":
			return evaluate(node.left, tags) || evaluate(node.right, tags);
		case "xor":
			return evaluate(node.left, tags) !== evaluate(node.right, tags);
		case "diff":
			return evaluate(node.left, tags) && !evaluate(node.right, tags);
	}
}

/**
 * Tokenizes, parses, and evaluates a tag expression against a tag set.
 * Throws on malformed expressions (empty, bad characters, syntax errors).
 */
export function matchTagExpr(expr: string, tags: ReadonlySet<string>): boolean {
	const tokens = tokenize(expr);
	if (tokens.length === 0) {
		throw new Error(errTagExprEmpty());
	}
	const node = new TagParser(tokens).parse();
	return evaluate(node, tags);
}
